core/src/main/java/com/linecorp/armeria/common/logging/ByteBufAggreatedPreviewer.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.logging;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.BiFunction;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaders;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufHolder;
+import io.netty.buffer.Unpooled;
+import io.netty.util.ReferenceCountUtil;
+
+abstract class ByteBufAggreatedPreviewer implements ContentPreviewer {
+
+    private static final ByteBuf[] BYTE_BUFS = new ByteBuf[0];
+    private int capacity;
+    private final List<ByteBuf> bufferList;
+    @Nullable
+    private HttpHeaders headers;
+    @Nullable
+    private String produced;
+    private int aggregatedLength;
+
+    ByteBufAggreatedPreviewer(int capacity) {
+        checkArgument(capacity >= 0, "capacity: %d (expected: >= 0)", capacity);
+        this.capacity = capacity;
+        bufferList = new ArrayList<>();
+    }
+
+    ByteBufAggreatedPreviewer() {
+        bufferList = new ArrayList<>();
+        capacity = 0;
+    }
+
+    void increaseCapacity(int delta) {
```

Comment:
It seems like it is called only once at max. How about replacing this with `setMaxAggregatedLength(int)`? For safety, we could add an assertion like: `assert maxAggregatedLength == 0 : "setMaxAggregatedLength() called more than once.";`